### PR TITLE
Line Rasterizer + ClipToGrid

### PR DESCRIPTION
**Pull request by fosskers**
This PR combines the work by @jpolchlo and @lossyrob in their PRs #2438 and #2424 respectively. Those two PRs are superseded by this one.

Once merged, this new functionality can be used directly by both Pointcloud and VectorPipe.

Closes #2438 
Closes #2424

### CODE REVIEW

File: vector-testkit/src/main/scala/geotrellis/vector/testkit/package.scala [hunk 1]
```diff
@@ -33,14 +33,19 @@ package object testkit {
     def matchPoint(p1: Point, p2: Point, tolerance: Double): Boolean =
       math.abs(p1.x - p2.x) <= tolerance && math.abs(p1.y - p2.y) <= tolerance
 
-    def matchLine(l1: Line, l2: Line, tolerance: Double): Boolean =
+    def matchLine(l1: Line, l2: Line, tolerance: Double): Boolean = {
+      println(l1.normalized)
+      println(l2.normalized)
```
Comment (lossyrob) [inline vector-testkit/src/main/scala/geotrellis/vector/testkit/package.scala]: Sorry, left some printlns in here

Comment (fosskers) [reply vector-testkit/src/main/scala/geotrellis/vector/testkit/package.scala]: Killing them with fire, one sec.

**Comment by fosskers** [comment]
One of the testing machines passed, but the other failed with:
```
java.lang.OutOfMemoryError: GC overhead limit exceeded
	at java.lang.Integer.valueOf(Integer.java:832)
	at scala.runtime.BoxesRunTime.boxToInteger(BoxesRunTime.java:65)
	at scala.Tuple2$mcII$sp._2(Tuple2.scala:20)
	at scala.Product2$class.productElement(Product2.scala:38)
	at scala.Tuple2.productElement(Tuple2.scala:20)
	at scala.util.hashing.MurmurHash3.productHash(MurmurHash3.scala:64)
	at scala.util.hashing.MurmurHash3$.productHash(MurmurHash3.scala:211)
	at scala.runtime.ScalaRunTime$._hashCode(ScalaRunTime.scala:168)
	at scala.Tuple2.hashCode(Tuple2.scala:20)
	at java.util.concurrent.ConcurrentHashMap.putVal(ConcurrentHashMap.java:1012)
	at java.util.concurrent.ConcurrentHashMap.put(ConcurrentHashMap.java:1006)
	at geotrellis.spark.tiling.MapKeyTransform$$anon$1.apply$mcVII$sp(MapKeyTransform.scala:145)
	at geotrellis.raster.rasterize.Rasterizer$.foreachCellInGridLineDouble(Rasterizer.scala:416)
	at geotrellis.raster.rasterize.Rasterizer$.foreachCellByLineStringDouble(Rasterizer.scala:339)
	at geotrellis.spark.tiling.MapKeyTransform$$anonfun$multiLineToKeys$1.apply(MapKeyTransform.scala:149)
	at geotrellis.spark.tiling.MapKeyTransform$$anonfun$multiLineToKeys$1.apply(MapKeyTransform.scala:149)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
	at geotrellis.spark.tiling.MapKeyTransform.multiLineToKeys(MapKeyTransform.scala:149)
	at geotrellis.spark.clip.ClipToGrid$.geotrellis$spark$clip$ClipToGrid$$clipGeom(ClipToGrid.scala:163)
	at geotrellis.spark.clip.ClipToGrid$$anonfun$apply$8.apply(ClipToGrid.scala:108)
	at geotrellis.spark.clip.ClipToGrid$$anonfun$apply$8.apply(ClipToGrid.scala:108)
...
```

**Comment by jpolchlo** [comment]
Do you have the geometry and raster extent of the failed call?

**Comment by fosskers** [comment]
@jpolchlo I suspect it's because of the resource-constrained travis machines. The other one didn't fail.

**Comment by fosskers** [comment]
I rebooted the CI build, we'll see what happens.

**Comment by echeipesh** [comment]
Not sure what is triggering ip-validation failure here, all involved are committers and have good git configurations, seems good.

**Pull request merged**